pkg/admission/permissionclaims/mutating_permission_claims.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package permissionclaims
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/admission"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1/permissionclaims"
+	kcpinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+)
+
+const (
+	byWorkspaceIndex = "mutating-permission-byWorkspace"
+	PluginName       = "apis.kcp.dev/PermissionClaims"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(configFile io.Reader) (admission.Interface, error) {
+		return NewMutatingPermissionClaims(), nil
+	})
+}
+
+type mutatingPermissionClaims struct {
+	*admission.Handler
+
+	apiBindingsIndexer cache.Indexer
+	bindingReady       func() bool
+}
+
+var _ admission.MutationInterface = &mutatingPermissionClaims{}
+
+// NewMutatingPermissionClaims will createa a mutating admission plugin that is responsible for labeling permission claims
+// For every creation request, we will determine the bindings in the workspace and if the object is claimed by an accepted
+// permission claim we will add the label.
+func NewMutatingPermissionClaims() admission.MutationInterface {
+	p := &mutatingPermissionClaims{}
+	p.Handler = admission.NewHandler(admission.Create)
+	p.SetReadyFunc(p.bindingReady)
+	return p
+}
+
+func (m *mutatingPermissionClaims) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	bindings, err := m.apiBindingsIndexer.ByIndex(byWorkspaceIndex, lcluster.String())
+	if err != nil {
+		return err
+	}
+
+	for _, b := range bindings {
+		binding, ok := b.(*apisv1alpha1.APIBinding)
+		if !ok {
+			return fmt.Errorf("expected a certain type")
+		}
+		for _, pc := range binding.Status.ObservedAcceptedPermissionClaims {
+			if pc.Group == a.GetKind().Group && pc.Resource == a.GetResource().Resource {
```

Comment:
how do we know the identity matches? We could store the real resource identity in the context somehow and then compare here. cc @ncdc

Reply:
updated